File: iOS/SmartPlanner/Domain/Tasks/CreateTaskUseCase.swift
//
//  CreateTaskUseCase.swift
//  SmartPlanner
//

import Foundation

struct CreateTaskInput {
    let title: String
    let details: String?
    let priority: TaskPriority
    let isFlagged: Bool
    let deadline: Date?
}

enum CreateTaskError: Error, Equatable {
    case invalidTitle
}

protocol CreateTaskUseCaseProtocol {
    func execute(_ input: CreateTaskInput) -> Result<Task, CreateTaskError>
}

struct CreateTaskUseCase: CreateTaskUseCaseProtocol {

    private let validator: TaskValidating
    private let repository: TaskRepositoryProtocol

    init(
        validator: TaskValidating = TaskValidator(),
        repository: TaskRepositoryProtocol = InMemoryTaskRepository.shared
    ) {
        self.validator = validator
        self.repository = repository
    }

    func execute(_ input: CreateTaskInput) -> Result<Task, CreateTaskError> {
        guard validator.isValidTitle(input.title) else {
            return .failure(.invalidTitle)
        }

        let task = Task(
            title: input.title.trimmingCharacters(in: .whitespacesAndNewlines),
            details: validator.sanitize(input.details),
            priority: input.priority,
            isFlagged: input.isFlagged,
            deadline: input.deadline
        )

        repository.save(task)
        return .success(task)
    }
}

File: Android/app/src/main/java/com/example/smartplannercompose/domain/tasks/CreateTaskUseCase.kt
package com.example.smartplannercompose.domain.tasks

import com.example.smartplannercompose.core.Clock
import com.example.smartplannercompose.core.SystemClock

data class CreateTaskInput(
    val title: String,
    val details: String?,
    val priority: TaskPriority,
    val isFlagged: Boolean,
    val deadlineMillis: Long?
)

sealed class CreateTaskError : Throwable() {
    data object InvalidTitle : CreateTaskError()
}

interface CreateTaskUseCase {
    fun execute(input: CreateTaskInput): Result<Task>
}

class DefaultCreateTaskUseCase(
    private val validator: TaskValidator = DefaultTaskValidator(),
    private val repository: TaskRepository,
    private val clock: Clock = SystemClock
) : CreateTaskUseCase {

    override fun execute(input: CreateTaskInput): Result<Task> {
        if (!validator.isValidTitle(input.title)) {
            return Result.failure(CreateTaskError.InvalidTitle)
        }
        val task = Task.create(
            title = input.title.trim(),
            details = validator.sanitize(input.details),
            priority = input.priority,
            isFlagged = input.isFlagged,
            deadlineMillis = input.deadlineMillis,
            createdAtMillis = clock.nowMillis()
        )
        repository.save(task)
        return Result.success(task)
    }
}
